### PR TITLE
reflatten: better control of flattening

### LINKED ISSUE #697: Flatten parameters by group
## 🚀 Feature

FSDP to offer the possibility to flatten parameters by group, for instance, to flatten all biases separately from the other weights.

## Motivation

Following issue https://github.com/facebookresearch/fairscale/issues/644 and the attempt in solving it in PR https://github.com/facebookresearch/fairscale/pull/692, providing a view of the parameters in the view is too low level and does not offer the best user experience as there are plenty of ways to trip and fall (see PR for the limitations: https://github.com/facebookresearch/fairscale/pull/692).

Following a discussion with @min-xu-ai, the best is to go back at the uses cases motivating https://github.com/facebookresearch/fairscale/issues/644:

1. be able to easily compute the weight and gradient norms when `flatten_parameters=True` needed for LARC like optimisers
2. be able to have separate LR, regularisation for each parameters (for instance, regularise only bias and not weight or the other way around)

This issue proposes to solve item 2.

## Workarounds

There is no workaround for the moment when `flatten_parameters=True`: having different wrappers for different modules does not offer the granularity required to flatten biases and weights of the `nn.Linear` layer separately for instance.

## Interested parties

CC: @min-xu-ai @myleott @prigoyal

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?
Fixes #697 .

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by anj-s** [comment]
@min-xu-ai  Is this still in progress?

**Comment by min-xu-ai** [comment]
this is on hold. I will try to find one of those "free" weeks. :-)

**Comment by anj-s** [comment]
> this is on hold. I will try to find one of those "free" weeks. :-)

Oh those elusive weeks...I've been trying to get hold of them myself :)
Reason I asked is I was trying to figure out if these changes you are working on will help code quality or will address a feature request. I was making a list of FSDP changes and I remembered this PR.
